p4-16/psa/PSA.mdk

```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
 drop all packets to an output port, that port will go idle.  The
 scheduling algorithm implementations are all built with a finite
 maximum packet scheduling rate.
+
+
+# Appendix: Multi-pipeline PSA devices
+
+The highest packet rate network devices today are ASICs running at a
+clock rate on the order of 1 to 2 GHz.  This discussion will assume 1
+GHz for the sake of a concrete numerical example, but everything
+discussed here scales linearly with the clock rate.
+
+It is common to design a portion of a network ASIC such that it can
+start processing a new packet once every clock cycle, and finish a
+packet every clock cycle.  The latency might be hundreds of clock
+cycles from starting a packet until it is complete.  P4 tables in such
+ASICs are typically implemented using logic such as TCAMs and SRAMs.
+TCAM designs can do 1 search per clock cycle.  The lowest area and
+power SRAMs can do 1 read or 1 write per clock cycle.
+
+While there are "multi ported" SRAM designs that can be read and/or
+written multiple times per clock cycle, these have a noticeable
+increase in area and power over the "singled ported" designs that are
+limited to 1 access per cycle.  If multi ported TCAM designs even
+exist, they are likely even a higher cost premium of single ported
+TCAMs than for SRAMs.  They typical way to achieve higher search rates
+is via parallelism, by creating multiple copies of the desired TCAM,
+which is a linear increase in area and power (at least).
+
+Due to these issues, if one wishes to create a switch ASIC that
+achieves a packet processing rate of N times the clock rate, e.g. 2
+billion packets per second in a 1 GHz ASIC, the most straightforward
+way to do this is to take advantage of the fact that packet processing
+is (mostly) an embarrasingly parallel
+problem[^EmbarrassinglyParallel], and create a device with N modules,
+each modules processing packets at 1 billion packets per second.
```

Comment:
"each modules" => each module. Do you think it might help to use the term pipeline instead of module? Or was your intent to cover both pipeline and NPU-style architectures?

Reply:
Yes, I will use pipeline throughout, trying to restrict myself on the first use to a footnote that the term should not be taken to mean that a hardware pipeline is the only kind of implementation possible.